docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+There is another restriction: in case shoot clusters are configured to be seed clusters this network range must not overlap with the "parent" seed cluster. If the parent seed cluster uses `192.168.0.0/22` the child seed cluster can for example use `192.168.4.0/22`. Grandchildren can however use grandparent IP address ranges.
+
+This slightly adds to the restrictions described in the current solution outline. In that the arbitrary chosen `192.168.123.0/24` range is resticted. For the purpose of this implementation we propose to extend that restriction to `192.168.128.0/17` range. Most of it would be reserved for "future use however".
+
+WireGuard can utilize the Linux kernel so that after initialization/configuration no user space processes are required. We propose to recommend the wireguard kernel module as the default solution for all seeds. For shoot clusters, the WireGuard kernel based approach is also recommended, but the user space solution should also work as we expect less traffic on the shoot side. We expect the userspace implementation to work on all operating systems supported by Gardener in case no kernel module is available.
+
+There is a user space implementation of WireGuard, which can be used on Linux distributions without the WireGuard kernel module. (WireGuard moved into the standard Linux kernel 5.6.) Our proposal can handle the kernel/user space switch transparently, i.e. we include the user space binaries and use them only when required. However, especially for the seed the kernel based solution might be more attractive. Garden Linux 318.4.0 supports WireGuard.
+
+We have looked at Ubuntu and SuSE chost:
+
+- SuSE chost does not provide the WireGuard kernel module and it is not installable via zypper. It should however be straightforward for SuSE to include that in their next release.
+- Ubuntu does not provide the kernel module either but it can be installed using `apt-get install wireguard`. With that it appears straightforward to provide an image with WireGuard pre-installed.
+
+On the seed, we add a WireGuard device to one node on the host network. For all other nodes on the seed, we adapt the routes accordingly to route traffic destined for the WireGuard network to our WireGuard node. The Kubernetes pods managing the WireGuard device and routes are only used for initial configuration and later reconfiguration. During runtime, they can restart without any impact on the operation of the WireGuard network as the WireGuard device is managed by the Linux kernel.
+
+The WireGuard endpoint on the seed is exposed via a load balancer. We propose to use kubelink to manage the WireGuard configuration/device on the seed. We consider the management of the WireGuard endpoint to be complex especially in error situations which is the reason for utilizing kubelink as there is already significant experience managing an endpoint.
```

Comment:
Exactly. Please explain what Kubelink is (btw, why is it still not in the `gardener` org?) and why it is needed. You say

> We consider the management of the WireGuard endpoint to be complex especially in error situations

Can you explain more and how kubelink helps?

Reply:
yes, we will add this and Uwe already mentioned that this should then be moved to the gardenr org. I do agree.

Reply:
> We consider the management of the WireGuard endpoint to be complex especially in error situations

Can you share more details? What makes it complex? What needs to be managed? What are the risk in such error situations and how kubelink helps?


> (btw, why is it still not in the gardener org?)

I do not see this as a hard requirement. k/k is also not in the gardener org :sunglasses:

Reply:
The management of the wireguard device becomes more complex as it is currently created on the host network, i.e. it is not automatically cleaned up in case the corresponding pod terminates. Therefore, additional steps are required for proper clean-up.
What needs to be managed? The wireguard device needs to be managed. Furthermore, the routes to the device (from other nodes) need to be managed.
kubelink helps in that it does the management of the wireguard device and the routes to it.